rs/crypto/secp256k1/src/lib.rs

```diff
@@ -533,6 +585,41 @@ impl PrivateKey {
         self.sign_message_with_bip340(message, &mut rng)
     }
 
+    /// BIP341 derivation
+    fn derive_bip341(&self, ttr: &[u8]) -> Result<Self, InvalidTaprootHash> {
+        let pk = self.public_key().serialize_sec1(true);
+
+        let t = bip341_hash(&pk[1..], ttr)?;
+        let pk_y_is_even = pk[0] == 0x02;
+
+        let z = if pk_y_is_even {
+            self.key.to_nonzero_scalar().as_ref() + t
+        } else {
+            self.key.to_nonzero_scalar().as_ref().negate() + t
+        };
+
+        let nz_ds = k256::NonZeroScalar::new(z).expect("Derivation always produces non-zero sum");
+
+        Ok(Self {
+            key: k256::SecretKey::from(nz_ds),
+        })
+    }
+
+    /// Sign a message with BIP340 Schnorr with Taproot derivation
+    pub fn sign_message_with_bip341<R: Rng + CryptoRng>(
+        &self,
+        message: &[u8],
+        rng: &mut R,
+        taproot_tree_hash: &[u8],
+    ) -> Result<[u8; 64], InvalidTaprootHash> {
+        if !taproot_tree_hash.is_empty() && taproot_tree_hash.len() != 32 {
+            return Err(InvalidTaprootHash::InvalidLength);
+        }
+
+        let dk = self.derive_bip341(taproot_tree_hash)?;
```

Comment:
There are still 4 occurrences of `dk`, while some were renamed to `tweaked_key`. Should we use one name consistently?

Reply:
Fixed